src/Orleans/Storage/OrleansRelationalExtensions.cs

```diff
@@ -0,0 +1,1506 @@
+        {
+            var query = OrleansRelationalConstants.GetConstant(storage.InvariantName, OrleansRelationalConstants.ActiveGatewaysQuery);
+            var ret = await storage.ReadAsync(query, command =>
+            {
+                var direction = ParameterDirection.Input;
+                var siloIdParameter = CreateDeploymentIdParameter(command, deploymentId, direction);
+                command.Parameters.Add(siloIdParameter);
+
+                var statusParameter = CreateStatusParameter(command, SiloStatus.Active, direction);
+                command.Parameters.Add(statusParameter);
+            }, (selector, _) =>
+            {
+                var ip = selector.GetValue<string>("Address");
+                var port = selector.GetValue<int>("ProxyPort");
+
+                return new IPEndPoint(IPAddress.Parse(ip), port).ToGatewayUri();
+            }).ConfigureAwait(continueOnCapturedContext: false);
+
+            return ret.ToList();
+        }
+
+
+        /// <summary>
+        /// Queries Orleans membership data.
+        /// </summary>
+        /// <param name="storage">The storage to use.</param>
+        /// <param name="deploymentId">The deployment for which to query data.</param>
+        /// <param name="key">Silo data used as parameters in the query.</param>
+        /// <returns>Membership table data.</returns>
+        internal static async Task<MembershipTableData> MembershipDataAsync(this IRelationalStorage storage, string deploymentId, SiloAddress key)
+        {
+            var query = OrleansRelationalConstants.GetConstant(storage.InvariantName, OrleansRelationalConstants.MembershipReadRowKey);
+            var ret = await storage.ReadAsync(query, command =>
+            {
+                var direction = ParameterDirection.Input;
+                var siloIdParameter = CreateDeploymentIdParameter(command, deploymentId, direction);
+                command.Parameters.Add(siloIdParameter);
+
+                var addressParameter = CreateAddressParameter(command, key.Endpoint.Address, direction);
+                command.Parameters.Add(addressParameter);
+
+                var portParameter = CreatePortParameter(command, key.Endpoint.Port, direction);
+                command.Parameters.Add(portParameter);
+
+                var generationParameter = CreateGenerationParameter(command, key.Generation, direction);
+                command.Parameters.Add(generationParameter);
+            }, (selector, _) =>
+            {
+                return CreateMembershipEntry(selector);
+            }).ConfigureAwait(continueOnCapturedContext: false);
+
+            //All the rows have the same Version (.Item3) and VersionETag (.Item4) information.
+            //If .Item1 of the first element has no data, the MembershipEntry collection is empty.
+            var membershipData = ret.First().Item1 == null ? Enumerable.Empty<Tuple<MembershipEntry, string>>() : ret.Select(i => Tuple.Create(i.Item1, i.Item2));
```

Comment:
Multiple enumerations of IEnumerable. Introduce variable for ret.First().

+ same pattern is several other places in this file.

[Found by ReSharper]